File: app/src/main/java/com/babestudios/companyinfouk/ui/search/SearchActivity.java
package com.babestudios.companyinfouk.ui.search;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.ProgressBar;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.babestudios.companyinfouk.CompaniesHouseApplication;
import com.babestudios.companyinfouk.R;
import com.babestudios.companyinfouk.data.DataManager;
import com.babestudios.companyinfouk.data.model.search.CompanySearchResult;
import com.babestudios.companyinfouk.data.model.search.SearchHistoryItem;
import com.babestudios.companyinfouk.ui.company.CompanyActivity;
import com.babestudios.companyinfouk.ui.favourites.FavouritesActivity;
import com.babestudios.companyinfouk.uiplugins.BaseActivityPlugin;
import com.babestudios.companyinfouk.utils.DividerItemDecoration;
import com.babestudios.companyinfouk.utils.DividerItemDecorationWithSubHeading;
import com.babestudios.companyinfouk.utils.EndlessRecyclerViewScrollListener;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.pascalwelsch.compositeandroid.activity.CompositeActivity;

import net.grandcentrix.thirtyinch.plugin.TiActivityPlugin;

import java.util.ArrayList;
import java.util.Arrays;

import javax.inject.Inject;
import javax.inject.Singleton;

import butterknife.Bind;
import butterknife.ButterKnife;

public class SearchActivity extends CompositeActivity implements SearchActivityView, SearchResultsAdapter.SearchResultsRecyclerViewClickListener,
		RecentSearchesResultsAdapter.RecentSearchesRecyclerViewClickListener {

	@Bind(R.id.toolbar)
	Toolbar toolbar;

	@Bind(R.id.recent_searches_recycler_view)
	RecyclerView recentSearchesRecyclerView;

	@Bind(R.id.search_recycler_view)
	RecyclerView searchRecyclerView;

	@Bind(R.id.fab)
	FloatingActionButton fab;

	private SearchResultsAdapter searchResultsAdapter;

	private RecentSearchesResultsAdapter recentSearchesResultsAdapter;

	@Bind(R.id.progressbar)
	ProgressBar progressbar;

	MenuItem searchMenuItem;

	@Inject
	SearchPresenter searchPresenter;

	private SearchPresenter.FilterState initialFilterState;

	TiActivityPlugin<SearchPresenter, SearchActivityView> searchActivityPlugin = new TiActivityPlugin<>(
			() -> {
				CompaniesHouseApplication.getInstance().getApplicationComponent().inject(SearchActivity.this);
				return searchPresenter;
			});

	BaseActivityPlugin baseActivityPlugin = new BaseActivityPlugin();

	public SearchActivity() {
		addPlugin(searchActivityPlugin);
		addPlugin(baseActivityPlugin);
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_search);
		baseActivityPlugin.logScreenView(this.getLocalClassName());

		ButterKnife.bind(this);
		if (toolbar != null) {
			setSupportActionBar(toolbar);
			getSupportActionBar().setDisplayHomeAsUpEnabled(false);
			toolbar.setNavigationOnClickListener(v -> onBackPressed());
		}
		createRecentSearchesRecyclerView();
		createSearchResultsRecyclerView();
		fab.setOnClickListener(view -> searchActivityPlugin.getPresenter().onFabClicked());
		performAnimations();
	}

	private void performAnimations() {
		fab.setTranslationY(2 * getResources().getDimensionPixelOffset(R.dimen.fab_size));
		fab.animate()
				.translationY(0)
				.setInterpolator(new DecelerateInterpolator())
				.setStartDelay(getResources().getInteger(R.integer.fab_move_in_start_delay))
				.setDuration(getResources().getInteger(R.integer.fab_move_in_duration))
				.start();
	}

	private void createRecentSearchesRecyclerView() {
		LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
		recentSearchesRecyclerView.setLayoutManager(linearLayoutManager);
		ArrayList<Integer> titlePositions = new ArrayList<>();
		titlePositions.add(0);
		recentSearchesRecyclerView.addItemDecoration(
				new DividerItemDecorationWithSubHeading(this, titlePositions));
	}

	private void createSearchResultsRecyclerView() {
		LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
		searchRecyclerView.setLayoutManager(linearLayoutManager);
		searchRecyclerView.addItemDecoration(
				new DividerItemDecoration(this));

		searchRecyclerView.addOnScrollListener(new EndlessRecyclerViewScrollListener(linearLayoutManager) {
			@Override
			public void onLoadMore(int page, int totalItemsCount) {
				searchActivityPlugin.getPresenter().searchLoadMore(page);
			}
		});
	}

	@Override
	public void showProgress() {
		progressbar.setVisibility(View.VISIBLE);
	}

	@Override
	public void hideProgress() {
		progressbar.setVisibility(View.GONE);
	}

	@Override
	public void showRecentSearches(SearchHistoryItem[] searchHistoryItems) {
		recentSearchesRecyclerView.setVisibility(View.VISIBLE);
		searchRecyclerView.setVisibility(View.GONE);
		if (recentSearchesRecyclerView.getAdapter() == null) {
			ArrayList<SearchHistoryItem> searchHistoryItemsList;
			if (searchHistoryItems != null) {
				searchHistoryItemsList = new ArrayList<>(Arrays.asList(searchHistoryItems));
			} else {
				searchHistoryItemsList = new ArrayList<>();
			}
			recentSearchesResultsAdapter = new RecentSearchesResultsAdapter(SearchActivity.this, searchHistoryItemsList);
			recentSearchesRecyclerView.setAdapter(recentSearchesResultsAdapter);
		}
	}

	@Override
	public void startCompanyActivity(String companyNumber, String companyName) {
		Intent startCompanyActivityIntent = new Intent(this, CompanyActivity.class);
		startCompanyActivityIntent.putExtra("companyNumber", companyNumber);
		startCompanyActivityIntent.putExtra("companyName", companyName);
		baseActivityPlugin.startActivityWithRightSlide(startCompanyActivityIntent);
	}

	/**
	 * @param companySearchResult
	 * @param isFromOnNext:       addItems on adapter should only be used for onLoadMore from onNext
	 * @param filterState
	 */
	@Override
	public void showCompanySearchResult(CompanySearchResult companySearchResult, boolean isFromOnNext, SearchPresenter.FilterState filterState) {
		recentSearchesRecyclerView.setVisibility(View.GONE);
		searchRecyclerView.setVisibility(View.VISIBLE);
		if (searchRecyclerView.getAdapter() == null) {
			searchResultsAdapter = new SearchResultsAdapter(SearchActivity.this, companySearchResult, filterState);
			searchRecyclerView.setAdapter(searchResultsAdapter);
		} else if (isFromOnNext) {
			searchResultsAdapter.addItems(companySearchResult, filterState);
		}
	}

	@Override
	public void clearSearchView() {
		if (searchMenuItem != null) {
			MenuItemCompat.collapseActionView(searchMenuItem);
		}
	}

	@Override
	public void refreshRecentSearchesAdapter(ArrayList<SearchHistoryItem> searchHistoryItems) {
		recentSearchesResultsAdapter.refreshData(searchHistoryItems);
	}

	@Override
	public void changeFabImage(SearchPresenter.FabImage type) {
		fab.animate()
				.translationY(6 * getResources().getDimensionPixelOffset(R.dimen.fab_size))
				.setStartDelay(100)
				.setInterpolator(new AccelerateInterpolator())
				.setDuration(getResources().getInteger(R.integer.fab_move_in_duration))
				.setListener(new AnimatorListenerAdapter() {
					@Override
					public void onAnimationEnd(Animator animation) {
						if (type == SearchPresenter.FabImage.FAB_IMAGE_RECENT_SEARCH_DELETE) {
							fab.setImageResource(R.drawable.delete_vector);
						} else if (type == SearchPresenter.FabImage.FAB_IMAGE_SEARCH_CLOSE) {
							fab.setImageResource(R.drawable.close_vector);
						}
						fab.animate()
								.translationY(0)
								.setInterpolator(new DecelerateInterpolator())
								.setStartDelay(100)
								.setDuration(getResources().getInteger(R.integer.fab_move_in_duration))
								.start();
					}
				})
				.start();
	}

	@Override
	public void showDeleteRecentSearchesDialog() {
		new AlertDialog.Builder(this)
				.setTitle(R.string.delete_recent_searches)
				.setMessage(R.string.are_you_sure_you_want_to_delete_all_recent_searches)
				.setPositiveButton(android.R.string.yes, (dialog, which) -> {
					searchActivityPlugin.getPresenter().clearAllRecentSearches();
				})
				.setNegativeButton(android.R.string.no, (dialog, which) -> {
					// do nothing
				})
				.show();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.search_menu, menu);

		//Catch the click on the search button on the soft keyboard and send the query to the presenter
		searchMenuItem = menu.findItem(R.id.action_search);
		SearchView searchView = (SearchView) MenuItemCompat.getActionView(searchMenuItem);
		TextView searchText = (TextView) searchView.findViewById(android.support.v7.appcompat.R.id.search_src_text);
		searchText.setOnEditorActionListener((textView, actionId, keyEvent) -> {
			if (actionId == EditorInfo.IME_ACTION_SEARCH) {
				searchRecyclerView.setAdapter(null);
				View view = this.getCurrentFocus();
				if (view != null) {
					view.clearFocus();
					InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
					imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
				}

				String queryText = searchView.getQuery().toString();
				logSearch(queryText);
				searchActivityPlugin.getPresenter().search(queryText);
				return true;
			}
			return false;
		});
		MenuItem item = menu.findItem(R.id.spinner);
		Spinner spinner = (Spinner) MenuItemCompat.getActionView(item);
		spinner.setBackgroundResource(0);
		spinner.setPadding(0, 0, getResources().getDimensionPixelOffset(R.dimen.view_margin), 0);
		SearchFilterAdapter adapter = new SearchFilterAdapter(SearchActivity.this, getResources().getStringArray(R.array.search_filter_options), true);
		spinner.setAdapter(adapter);
		if (initialFilterState != null) {
			spinner.setSelection(initialFilterState.ordinal());
			initialFilterState = null;
		}
		spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
				searchActivityPlugin.getPresenter().setFilterState(position);
			}

			@Override
			public void onNothingSelected(AdapterView<?> parent) {

			}
		});
		return true;
	}

	private void logSearch(String queryText) {
		Bundle bundle = new Bundle();
		bundle.putString(FirebaseAnalytics.Param.SEARCH_TERM, queryText);
		CompaniesHouseApplication.getInstance().getFirebaseAnalytics().logEvent(FirebaseAnalytics.Event.SEARCH, bundle);
	}

	@Override
	public void setInitialFilterState(SearchPresenter.FilterState filterState) {
		this.initialFilterState = filterState;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_favourites:
				Intent intent = new Intent(this, FavouritesActivity.class);
				baseActivityPlugin.startActivityWithRightSlide(intent);
				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	@Override
	public void searchResultItemClicked(View v, int position, String companyName, String companyNumber) {
		searchActivityPlugin.getPresenter().getCompany(companyName, companyNumber);
	}

	@Override
	public void recentSearchesResultItemClicked(View v, int position, String companyName, String companyNumber) {
		searchActivityPlugin.getPresenter().getCompany(companyName, companyNumber);
	}

	@Override
	public void setFilterOnAdapter(SearchPresenter.FilterState filterState) {
		searchResultsAdapter.setFilterOnAdapter(filterState);
	}

	@Override
	public void showError() {
		Toast.makeText(this, "Error", Toast.LENGTH_LONG).show();
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/search/SearchActivity.kt
package com.babestudios.companyinfouk.ui.search

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.SearchView
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.*
import butterknife.BindView
import butterknife.ButterKnife
import com.babestudios.base.view.DividerItemDecoration
import com.babestudios.base.view.DividerItemDecorationWithSubHeading
import com.babestudios.base.view.EndlessRecyclerViewScrollListener
import com.babestudios.companyinfouk.CompaniesHouseApplication
import com.babestudios.companyinfouk.R
import com.babestudios.companyinfouk.data.model.search.CompanySearchResult
import com.babestudios.companyinfouk.data.model.search.SearchHistoryItem
import com.babestudios.companyinfouk.ui.company.createCompanyIntent
import com.babestudios.companyinfouk.ui.favourites.FavouritesActivity
import com.babestudios.companyinfouk.ui.privacy.PrivacyActivity
import com.babestudios.companyinfouk.uiplugins.BaseActivityPlugin
import com.google.firebase.analytics.FirebaseAnalytics
import com.pascalwelsch.compositeandroid.activity.CompositeActivity
import net.grandcentrix.thirtyinch.plugin.TiActivityPlugin
import java.util.*
import javax.inject.Inject

class SearchActivity : CompositeActivity(), SearchActivityView, SearchResultsAdapter.SearchResultsRecyclerViewClickListener, RecentSearchesResultsAdapter.RecentSearchesRecyclerViewClickListener {

	@JvmField
	@BindView(R.id.toolbar)
	internal var toolbar: Toolbar? = null

	@JvmField
	@BindView(R.id.recent_searches_recycler_view)
	internal var recentSearchesRecyclerView: RecyclerView? = null

	@JvmField
	@BindView(R.id.search_recycler_view)
	internal var searchRecyclerView: RecyclerView? = null

	@JvmField
	@BindView(R.id.fab)
	internal var fab: FloatingActionButton? = null

	private var searchResultsAdapter: SearchResultsAdapter? = null

	private var recentSearchesResultsAdapter: RecentSearchesResultsAdapter? = null

	@JvmField
	@BindView(R.id.progressbar)
	internal var progressbar: ProgressBar? = null

	private var searchMenuItem: MenuItem? = null

	@Inject
	internal lateinit var searchPresenter: SearchPresenter

	private var initialFilterState: SearchPresenter.FilterState? = null

	internal var searchActivityPlugin = TiActivityPlugin<SearchPresenter, SearchActivityView> {
		CompaniesHouseApplication.instance.applicationComponent.inject(this@SearchActivity)
		searchPresenter
	}

	internal var baseActivityPlugin = BaseActivityPlugin()

	init {
		addPlugin(searchActivityPlugin)
		addPlugin(baseActivityPlugin)
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_search)
		baseActivityPlugin.logScreenView(this.localClassName)

		ButterKnife.bind(this)
		if (toolbar != null) {
			setSupportActionBar(toolbar)
			supportActionBar?.setDisplayHomeAsUpEnabled(false)
			toolbar?.setNavigationOnClickListener { onBackPressed() }
		}
		createRecentSearchesRecyclerView()
		createSearchResultsRecyclerView()
		fab?.setOnClickListener { searchActivityPlugin.presenter.onFabClicked() }
		performAnimations()
	}

	private fun performAnimations() {
		fab?.let {
			it.translationY = (2 * resources.getDimensionPixelOffset(R.dimen.fab_size)).toFloat()
			it.animate()
					.translationY(0f)
					.setInterpolator(DecelerateInterpolator())
					.setStartDelay(resources.getInteger(R.integer.fab_move_in_start_delay).toLong())
					.setDuration(resources.getInteger(R.integer.fab_move_in_duration).toLong())
					.start()
		}
	}

	private fun createRecentSearchesRecyclerView() {
		val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
		recentSearchesRecyclerView?.layoutManager = linearLayoutManager
		val titlePositions = ArrayList<Int>()
		titlePositions.add(0)
		recentSearchesRecyclerView?.addItemDecoration(
				DividerItemDecorationWithSubHeading(this, titlePositions))
	}

	private fun createSearchResultsRecyclerView() {
		val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
		searchRecyclerView?.layoutManager = linearLayoutManager
		searchRecyclerView?.addItemDecoration(
				DividerItemDecoration(this))

		searchRecyclerView?.addOnScrollListener(object : EndlessRecyclerViewScrollListener(linearLayoutManager) {
			override fun onLoadMore(page: Int, totalItemsCount: Int) {
				searchActivityPlugin.presenter.searchLoadMore(page)
			}
		})
	}

	override fun showProgress() {
		progressbar?.visibility = View.VISIBLE
	}

	override fun hideProgress() {
		progressbar?.visibility = View.GONE
	}

	override fun showRecentSearches(searchHistoryItems: ArrayList<SearchHistoryItem>) {
		recentSearchesRecyclerView?.visibility = View.VISIBLE
		searchRecyclerView?.visibility = View.GONE
		if (recentSearchesRecyclerView?.adapter == null) {
			recentSearchesResultsAdapter = RecentSearchesResultsAdapter(this@SearchActivity, searchHistoryItems)
			recentSearchesRecyclerView?.adapter = recentSearchesResultsAdapter
		}
	}

	override fun refreshRecentSearchesAdapter(searchHistoryItems: ArrayList<SearchHistoryItem>) {
		if (recentSearchesRecyclerView?.adapter == null) {
			recentSearchesResultsAdapter = RecentSearchesResultsAdapter(this@SearchActivity, searchHistoryItems)
			recentSearchesRecyclerView?.adapter = recentSearchesResultsAdapter
		} else {
			recentSearchesResultsAdapter?.refreshData(searchHistoryItems)
		}
	}

	override fun startCompanyActivity(companyNumber: String, companyName: String) {
		baseActivityPlugin.startActivityWithRightSlide(createCompanyIntent(companyNumber, companyName))
	}

	/**
	 * @param companySearchResult
	 * @param isFromOnNext:       addItems on adapter should only be used for onLoadMore from onNext
	 * @param filterState
	 */
	override fun showCompanySearchResult(companySearchResult: CompanySearchResult, isFromOnNext: Boolean, filterState: SearchPresenter.FilterState) {
		recentSearchesRecyclerView?.visibility = View.GONE
		searchRecyclerView?.visibility = View.VISIBLE
		if (searchRecyclerView?.adapter == null) {
			searchResultsAdapter = SearchResultsAdapter(this@SearchActivity, companySearchResult, filterState)
			searchRecyclerView?.adapter = searchResultsAdapter
		} else if (isFromOnNext) {
			searchResultsAdapter?.addItems(companySearchResult, filterState)
		}
	}

	override fun clearSearchView() {
		searchMenuItem?.collapseActionView()
	}

	override fun changeFabImage(type: SearchPresenter.FabImage) {
		fab?.also {
			it.animate()
					.translationY((6 * resources.getDimensionPixelOffset(R.dimen.fab_size)).toFloat())
					.setStartDelay(100)
					.setInterpolator(AccelerateInterpolator())
					.setDuration(resources.getInteger(R.integer.fab_move_in_duration).toLong())
					.setListener(object : AnimatorListenerAdapter() {
						override fun onAnimationEnd(animation: Animator) {
							if (type === SearchPresenter.FabImage.FAB_IMAGE_RECENT_SEARCH_DELETE) {
								it.setImageResource(R.drawable.ic_delete)
							} else if (type === SearchPresenter.FabImage.FAB_IMAGE_SEARCH_CLOSE) {
								it.setImageResource(R.drawable.ic_close)
							}
							it.animate()
									.translationY(0f)
									.setInterpolator(DecelerateInterpolator())
									.setStartDelay(100)
									.setDuration(resources.getInteger(R.integer.fab_move_in_duration).toLong())
									.start()
						}
					})
					.start()
		}
	}

	override fun showDeleteRecentSearchesDialog() {
		AlertDialog.Builder(this)
				.setTitle(R.string.delete_recent_searches)
				.setMessage(R.string.are_you_sure_you_want_to_delete_all_recent_searches)
				.setPositiveButton(android.R.string.yes) { _, _ -> searchActivityPlugin.presenter.clearAllRecentSearches() }
				.setNegativeButton(android.R.string.no) { _, _ ->
					// do nothing
				}
				.show()
	}

	override fun onCreateOptionsMenu(menu: Menu): Boolean {
		menuInflater.inflate(R.menu.search_menu, menu)

		//Catch the click on the search button on the soft keyboard and send the query to the presenter
		searchMenuItem = menu.findItem(R.id.action_search)
		val searchView = searchMenuItem?.actionView as SearchView
		val searchText = searchView.findViewById<View>(android.support.v7.appcompat.R.id.search_src_text) as TextView
		searchText.setOnEditorActionListener { _, actionId, _ ->
			if (actionId == EditorInfo.IME_ACTION_SEARCH) {
				searchRecyclerView?.adapter = null
				val view = this.currentFocus
				if (view != null) {
					view.clearFocus()
					val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
					imm.hideSoftInputFromWindow(view.windowToken, 0)
				}

				val queryText = searchView.query.toString()
				logSearch(queryText)
				searchActivityPlugin.presenter.search(queryText)
				return@setOnEditorActionListener true
			}
			false
		}
		val item = menu.findItem(R.id.spinner)
		val spinner = item.actionView as Spinner
		spinner.setBackgroundResource(0)
		spinner.setPadding(0, 0, resources.getDimensionPixelOffset(R.dimen.view_margin), 0)
		val adapter = SearchFilterAdapter(this@SearchActivity, resources.getStringArray(R.array.search_filter_options), true)
		spinner.adapter = adapter
		initialFilterState?.let {
			spinner.setSelection(it.ordinal)
			initialFilterState = null
		}
		spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
			override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
				searchActivityPlugin.presenter.setFilterState(position)
			}

			override fun onNothingSelected(parent: AdapterView<*>) {

			}
		}
		return true
	}

	private fun logSearch(queryText: String) {
		val bundle = Bundle()
		bundle.putString(FirebaseAnalytics.Param.SEARCH_TERM, queryText)
		CompaniesHouseApplication.instance.firebaseAnalytics?.logEvent(FirebaseAnalytics.Event.SEARCH, bundle)
	}

	override fun setInitialFilterState(filterState: SearchPresenter.FilterState) {
		this.initialFilterState = filterState
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		val intent: Intent
		return when (item.itemId) {
			R.id.action_favourites -> {
				intent = Intent(this, FavouritesActivity::class.java)
				baseActivityPlugin.startActivityWithRightSlide(intent)
				true
			}
			R.id.action_privacy -> {
				intent = Intent(this, PrivacyActivity::class.java)
				baseActivityPlugin.startActivityWithRightSlide(intent)
				true
			}
			else -> super.onOptionsItemSelected(item)
		}
	}

	override fun searchResultItemClicked(v: View, position: Int, companyName: String, companyNumber: String) {
		searchActivityPlugin.presenter.getCompany(companyName, companyNumber)
	}

	override fun recentSearchesResultItemClicked(v: View, position: Int, companyName: String, companyNumber: String) {
		searchActivityPlugin.presenter.getCompany(companyName, companyNumber)
	}

	override fun setFilterOnAdapter(filterState: SearchPresenter.FilterState) {
		searchResultsAdapter?.setFilterOnAdapter(filterState)
	}

	override fun showError() {
		Toast.makeText(this, R.string.could_not_complete_search, Toast.LENGTH_LONG).show()
	}
}
